Sysupgrade backup fixes

luizluca opened the pull request:
This patch series fixes some problems with sysupgrade backup.
I divided it in three steps as independent as possible from each other.
I think it is better to understand the change this way. However, I can squash
all or any one of them if needed.

These patches should also be backported to 19.07.

luizluca committed:
base-files: sysupgrade: do not backup files hidden by mount

sysupgrade '-c' or '-o' reads /overlay to save changed files.
However, some of those files might be hidden under a mount. This
can happen when services that stores files in an external device
runs when that device is not mounted, creating files in internal
flash. When backup is done after the device is again mounted, the
backup would either save the file from the external device (when it
exists there) or tar will fail to read the hidden file and backup
fails.

This change skip from backups files present in /overlay but hidden
by mounted dirs. It was tested with mount directories that uses
spaces and special characters like:

 mkdir "/etc/A\ \"!\$HOME' a*s?d[123] "
 touch "/etc/A\\ \"!\$HOME' a*s?d[123] "/hidden_file
 mount -t tmpfs none "/etc/A\\ \"!\$HOME' a*s?d[123] "

Signed-off-by: Luiz Angelo Daros de Luca <luizluca@gmail.com>

luizluca committed:
base-files: sysupgrade: better deal with errors

Instead of discarding tar stderr, it is better to remove
first slash from filelist and chdir / (-C /).

Fix cleanup when tar fails, removing files and umounting
temporary backup overlay directory.

Check for missing /overlay/upper and abort backup on filelist
errors.

Signed-off-by: Luiz Angelo Daros de Luca <luizluca@gmail.com>

jow- commented on package/base-files/files/sbin/sysupgrade:
```diff
@@ -118,25 +118,40 @@ list_conffiles() {
-	# Cannot handle spaces in filenames - but opkg cannot either...
-	list_conffiles | while read file csum; do
+	list_conffiles | while read -r csum_file; do
+		file="$(echo "$csum_file" | cut -d" " -f3-)"
```
Why is this code change needed? Why that inefficient `cut` usage? What is wrong with `read file csum` ?

luizluca replied:
> Why is this code change needed?

Current code does not support space in filename. I'm testing with an extreme case but any space at all would break sysupgrade like "/etc/my system/conf":

```
mkdir '/etc/A\ "!$HOME'"'"'  a*s?d[123] /'
touch '/etc/A\ "!$HOME'"'"'  a*s?d[123] /test_file'
echo '/etc/A\ "!$HOME'"'"'  a*s?d[123] /test_file' >>/etc/sysupgrade.conf
sysupgrade -l | grep test_file
```

I swapped file csum order to guarantee that the first matched "field separator" is not part of path. I used double space as "field separator" to match what sha256sum expects, so I do not need to build it later.

> What is wrong with `read file csum` ?

At least:
1. files will have only the first part of the filepath until it match any IFS. csum will have the remaining line.
2. any valid IFS will be converted and merged in a single space (tab, multiple spaces)
3. It is missing '-r', so \ is consumed by read

> Why that inefficient `cut` usage?

You are right. I overlooked it here. I pushed a new version that uses variable expansion "${csum_file#*  }", which is 300x faster.

adrianschmutzler commented on package/base-files/files/sbin/sysupgrade:
```diff
@@ -60,7 +60,7 @@ export INSTALLED_PACKAGES=${ETCBACKUP_DIR}/installed_packages.txt
 IMAGE="$1"
 
-[ -z "$IMAGE" -a -z "$NEED_IMAGE" -a $CONF_BACKUP_LIST -eq 0 -o $HELP -gt 0 ] && {
+[ -z "$IMAGE" ] && [ -z "$NEED_IMAGE" ] && [ "$CONF_BACKUP_LIST" -eq 0 ] || [ "$HELP" -gt 0 ] && {
```
I think you should separate the cosmetic changes (like this one) from the "functional" changes (space in filename) and make separate commits for them.

luizluca replied:
The cosmetic changes were so tittle that I kept it together with the shellcheck fixes. However, should label that commit on the most important aspect (the files with space fix) and not the minor issue (shellchecks).

Anyway, I splitted the commit.

adrianschmutzler replied:
I don't think it's a good idea to have strings in integer comparison.

luizluca committed:
base-files: sysupgrade: deal with spaces in path

Fix backup of files with spaces in path.
Fix upgrade with config file (-f) when config tar has spaces in path.

Signed-off-by: Luiz Angelo Daros de Luca <luizluca@gmail.com>

luizluca committed:
base-files: sysupgrade: shellcheck fixes

Signed-off-by: Luiz Angelo Daros de Luca <luizluca@gmail.com>

adrianschmutzler commented on package/base-files/files/sbin/sysupgrade:
```diff
@@ -101,7 +101,7 @@ EOF
 }
 
-[ -n "$IMAGE" -a -n "$NEED_IMAGE" ] && {
+[ -n "$IMAGE" ] && [ -n "$NEED_IMAGE" ] && {
```
This might change operator precedence.

adrianschmutzler commented:
Since nobody seems interested, consider closing the PR and submitting to the openwrt-devel list, where you might meet more interested people.

luizluca commented:
Thanks @adrianschmutzler , I even created a second PR #3045 with a subset.
